### PR TITLE
feat(core-data): implement Get Event by Id V2 API

### PR DESCRIPTION
**Pull request by cloudxxx8**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit subject follows the [Conventional Commits spec](https://github.com/zeke/semantic-pull-requests)
- [x] The commit message follows the [EdgeX Contributor Guide](https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] `make test` has completed successfully

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: Fix https://github.com/edgexfoundry/edgex-go/issues/2670

## What is the new behavior?
Implement Get Event by Id V2 API with unit test

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by cloudxxx8** [comment]
there is still one small unit test problem, but it is ready to be reviewed

**Comment by cloudxxx8** [comment]
The unit test problem has been resolved

File: internal/core/data/v2/controller/http/event_test.go [hunk 1]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+			recorder := httptest.NewRecorder()
+			handler := http.HandlerFunc(ec.EventById)
+			handler.ServeHTTP(recorder, req)
+
+			var actualResponse responseDTO.EventResponse
+			err = json.Unmarshal(recorder.Body.Bytes(), &actualResponse)
+			require.NoError(t, err)
+
+			if testCase.ErrorExpected {
+				assert.Equal(t, contractsV2.ApiVersion, actualResponse.ApiVersion, "API Version not as expected")
```
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event_test.go:282]: Add require.Error() before asserts

File: internal/core/data/v2/controller/http/event_test.go [hunk 2]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+				assert.Equal(t, testCase.ExpectedStatusCode, recorder.Result().StatusCode, "HTTP status code not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, int(actualResponse.StatusCode), "Response status code not as expected")
+			} else {
+				assert.Equal(t, contractsV2.ApiVersion, actualResponse.ApiVersion, "API Version not as expected")
```
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event_test.go:286]: Add require.NoError() before asserts

File: internal/core/data/v2/controller/http/event_test.go [hunk 3]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+			handler.ServeHTTP(recorder, req)
+
+			var actualResponse responseDTO.EventResponse
+			err = json.Unmarshal(recorder.Body.Bytes(), &actualResponse)
+			require.NoError(t, err)
+
+			if testCase.ErrorExpected {
+				assert.Equal(t, contractsV2.ApiVersion, actualResponse.ApiVersion, "API Version not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, recorder.Result().StatusCode, "HTTP status code not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, int(actualResponse.StatusCode), "Response status code not as expected")
```
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event_test.go:284]: Should also be checking the `Message` contains the error message.

File: internal/core/data/v2/controller/http/event_test.go [hunk 4]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+			} else {
+				assert.Equal(t, contractsV2.ApiVersion, actualResponse.ApiVersion, "API Version not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, recorder.Result().StatusCode, "HTTP status code not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, int(actualResponse.StatusCode), "Response status code not as expected")
+				assert.Equal(t, expectedMessage, actualResponse.Message, "Message not as expected")
```
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event_test.go:289]: We shouldn't be setting the Message on success.

File: internal/core/data/v2/controller/http/event_test.go [hunk 5]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+			}
```
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event_test.go:290]: What about verifying expected event was returned?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/data/v2/application/event.go [hunk 6]
```diff
@@ -91,3 +92,24 @@ func putEventOnQueue(evt dtos.Event, ctx context.Context, dic *di.Container) {
...
 	}
 }
+
+func EventById(id string, dic *di.Container) (dtos.Event, errors.EdgeX) {
+	if id == "" {
+		return dtos.Event{}, errors.NewCommonEdgeX(errors.KindInvalidId, "id is empty", nil)
+	} else {
+		_, err := uuid.Parse(id)
+		if err != nil {
+			return dtos.Event{}, errors.NewCommonEdgeX(errors.KindInvalidId, "uuid parsing failed", err)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/application/event.go:102]: modified

File: internal/core/data/v2/application/event_test.go [hunk 7]
```diff
@@ -140,3 +75,103 @@ func buildReadings() []models.Reading {
...
+		{"Invalid - EventId is not an UUID", invalidEventId, true, errors.KindInvalidId, http.StatusBadRequest},
+		{"Invalid - Event doesn't exist", notFoundEventId, true, errors.KindEntityDoesNotExist, http.StatusNotFound},
+	}
+
+	for _, testCase := range tests {
+		t.Run(testCase.Name, func(t *testing.T) {
+			evt, err := EventById(testCase.EventId, dic)
+
+			if testCase.ErrorExpected {
+				assert.NotEmpty(t, err.Error(), "Error message is empty")
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/application/event_test.go:169]: yes, thanks for the tip.  modified.

File: internal/core/data/v2/application/event_test.go [hunk 8]
```diff
@@ -140,3 +75,103 @@ func buildReadings() []models.Reading {
...
+				assert.Equal(t, testCase.ExpectedErrKind, errors.Kind(err), "Error kind not as expected")
+				assert.Equal(t, testCase.ExpectedStatusCode, err.Code(), "Error code not as expected")
+			} else {
+				assert.Equal(t, testCase.EventId, evt.ID, "Event Id not as expected")
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/application/event_test.go:173]: yes, thanks for the tip.  modified.

File: internal/core/data/v2/controller/http/const_test.go [hunk 9]
```diff
@@ -8,10 +8,14 @@ package http
...
+	TestPushedTime         = 1600666231295
+	TestCreatedTime        = 1600666214495
+	TestOriginTime         = 1600666185705354000
 	TestDeviceResourceName = "TestDeviceResourceName"
 
 	TestReadingValue           = "45"
 	TestBinaryReadingMediaType = "File"
 	TestReadingBinaryValue     = "testbinarydata"
+
+	InexistentEventID = "8ad33474-fbc5-11ea-adc1-0242ac120002"
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/const_test.go:20]: modified

File: internal/core/data/v2/controller/http/event.go [hunk 10]
```diff
@@ -57,25 +61,57 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
...
-	w.Header().Set(clients.ContentType, clients.ContentTypeJSON)
 	w.Header().Set(clients.CorrelationHeader, correlation.FromContext(ctx))
 	w.WriteHeader(http.StatusMultiStatus)
 	pkg.Encode(addResponses, w, lc)
 }
+
+func (ec *EventController) EventById(w http.ResponseWriter, r *http.Request) {
+	if r.Body != nil {
+		defer func() { _ = r.Body.Close() }()
+	}
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event.go:89]: removed

File: internal/core/data/v2/controller/http/event.go [hunk 11]
```diff
@@ -57,25 +61,57 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
...
+		lc.Debug(err.DebugMessages())
+		eventResponse = responseDTO.NewEventResponse("", err.Message(), err.Code(), e)
+		w.WriteHeader(err.Code())
+	} else {
+		eventResponse = responseDTO.NewEventResponse("", "Get event by id successfully", http.StatusOK, e)
+		w.WriteHeader(http.StatusOK)
+	}
+
+	w.Header().Set(clients.CorrelationHeader, correlation.FromContext(ctx))
+	pkg.Encode(eventResponse, w, lc)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event.go:116]: it's the package and function from V1.  should I create a new one for V2?

File: internal/core/data/v2/controller/http/event_test.go [hunk 12]
```diff
@@ -203,3 +228,66 @@ func TestAddEvent(t *testing.T) {
...
+
+	dic := mocks.NewMockDIC()
+	dic.Update(di.ServiceConstructorMap{
+		v2DataContainer.DBClientInterfaceName: func(get di.Get) interface{} {
+			return dbClientMock
+		},
+	})
+
+	ec := NewEventController(dic)
+	assert.NotNil(t, ec)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:252]: you are right.  removed

File: internal/core/data/v2/controller/http/event_test.go (hunk 1 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:282]: not really, the err here is the json parsing error.  In controller, there is no real error returned, and the error message is stated in the response with error http status code.

File: internal/core/data/v2/controller/http/event_test.go (hunk 2 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:286]: we have require.NoError() in front of the `if` block

File: internal/core/data/v2/controller/http/event_test.go (hunk 5 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:290]: modified

File: internal/core/data/v2/controller/http/event_test.go (hunk 4 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:289]: modified

File: internal/core/data/v2/controller/http/event_test.go (hunk 3 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:284]: ok, but this change will fail the unit test until https://github.com/edgexfoundry/go-mod-core-contracts/pull/291 merged

File: internal/core/data/v2/controller/http/event.go [hunk 13]
```diff
@@ -57,25 +61,57 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
...
+
+	// Get the event
+	e, err := application.EventById(id, ec.dic)
+	if err != nil {
+		lc.Error(err.Error())
+		lc.Debug(err.DebugMessages())
+		eventResponse = responseDTO.NewEventResponse("", err.Message(), err.Code(), e)
+		w.WriteHeader(err.Code())
+	} else {
+		eventResponse = responseDTO.NewEventResponse("", "Get event by id successfully", http.StatusOK, e)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event.go:111]: set it as empty string

File: internal/core/data/v2/controller/http/event_test.go (hunk 2 above)
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event_test.go:282]: modified after the error response becomes base response

File: internal/core/data/v2/controller/http/event.go [hunk 14]
```diff
@@ -57,25 +61,57 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
...
+	// URL parameters
+	vars := mux.Vars(r)
+	id := vars[v2.Id]
+
+	// Get the event
+	e, err := application.EventById(id, ec.dic)
+	if err != nil {
+		lc.Error(err.Error())
+		lc.Debug(err.DebugMessages())
+		eventResponse = responseDTO.NewEventResponse("", err.Message(), err.Code(), e)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/data/v2/controller/http/event.go]: modified

File: internal/core/data/v2/controller/http/event_test.go (hunk 1 above)
Comment (lenny-intel) [reply internal/core/data/v2/controller/http/event_test.go:275]: Ok, see it now.

File: internal/core/data/v2/controller/http/event.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/core/data/v2/controller/http/event.go:111]: Naw, just add comment so it is clear that it is also doing the send.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Comment by cloudxxx8** [comment]
@lenny-intel I look into the function, and there is no real "send" action here

```
func Encode(i interface{}, w http.ResponseWriter, LoggingClient logger.LoggingClient) {
	w.Header().Add(clients.ContentType, clients.ContentTypeJSON)

	enc := json.NewEncoder(w)
	err := enc.Encode(i)
	// Problems encoding
	if err != nil {
		LoggingClient.Error("Error encoding the data: " + err.Error())
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
}
```

**Comment by cloudxxx8** [comment]
recheck

**Comment by cloudxxx8** [comment]
recheck

**Comment by lenny-intel** [comment]
> 	enc := json.NewEncoder(w)

It is obscured by this line. Which is using the ResponseWriter to encode which sends the data.

**Comment by cloudxxx8** [comment]
> > enc := json.NewEncoder(w)
> 
> @cloudxxx8 , It is obscured by this line. Which is using the ResponseWriter to encode which sends the data.

@lenny-intel I added comment.  please help check again.  thanks.

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**